lib/setopt.c

```diff
@@ -77,6 +77,42 @@ CURLcode Curl_setstropt(char **charp, const char *s)
   return CURLE_OK;
 }
 
+CURLcode Curl_setblobopt(struct curl_blob**blobp,
+                         const struct curl_blob* st_blob)
+{
+  /* Release the previous storage at `charp' and replace by a dynamic storage
+     copy of `sblob'.
+     Return CURLE_OK, CURLE_BAD_FUNCTION_ARGUMENT or CURLE_OUT_OF_MEMORY. */
+
+  Curl_safefree(*blobp);
+
+  if(st_blob) {
+    /* sblob contain struct curl_blob data, created with
+        * curl_init_blob_*. This is not ordinary string but a
+        * blob memory block */
+
+    bool dup = (st_blob->flags) & CURL_BLOB_DUPFLAG_COPYMASK;
+    struct curl_blob* st_blob_dup;
+    st_blob_dup = (struct curl_blob*)malloc(sizeof(struct curl_blob) +
+                                            (dup ? st_blob->len : 0));
+    if(!st_blob_dup)
+        return CURLE_OUT_OF_MEMORY;
+    *st_blob_dup = *st_blob;
+    if(dup) {
+        /* dupflag is set by curl_init_blob_dup. We need copy the data,
+        *   app don't need keep data pointer valid after setopt */
+        st_blob_dup->data = (void *)(((char *)st_blob_dup) +
+                                sizeof(struct curl_blob));
+        memcpy(st_blob_dup->data, st_blob->data, st_blob->len);
+    }
```

Comment:
Looking at this logic I need to come back and question this. What's the point with the dup flag again? We provide the cert in memory from the beginning. What's the use case for telling libcurl to duplicate that?

Reply:
This is an important point.
Blob can be used in the future for bigger buffer than a small certificate (by example, full CACERT)

In CURLOPT_SSLCERT (and all other string) by example,  documentation say:
"The application does not have to keep the string around after setting this option. "
This is because string are duplicated. So used can discard the pointer.

There is one exception : CURLOPT_POSTFIELDS
documentation say "The data pointed to is NOT copied by the library"

For blob, duplicate is like string : ""The application does not have to keep". Without duplication, application need keep the buffer.
no duplicate is like POSTFIELD
This is like NSData objet in MacOSX : there is Init - bytes and Init - bytesNoCopy

Reply:
Do **you** have a use case where you want libcurl to duplicate the cert?

Reply:
When I made my custom version of curl with this feature, I wanted liburl duplicate the cert.
I got the certificate from another enrypted source, I immedialty create the curl objet (which started download on another thread), then I can discard the buffer without maintain the lifecycle.

Like option CURLOPT_COPYPOSTFIELDS which is a good thing

Reply:
Fair enough. Let's keep it. Let's instead work a little on making this function easy to follow.

Reply:
thank you !-)